src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD109AvoidAssertInAsyncMethodsAnalyzer.cs

```diff
@@ -49,43 +44,26 @@ public override void Initialize(AnalysisContext context)
             context.RegisterCompilationStartAction(ctxt =>
             {
                 var mainThreadAssertingMethods = CommonInterest.ReadMethods(ctxt.Options, CommonInterest.FileNamePatternForMethodsThatAssertMainThread, ctxt.CancellationToken).ToImmutableArray();
-                ctxt.RegisterSyntaxNodeAction(Utils.DebuggableWrapper(c => this.AnalyzeInvocation(c, mainThreadAssertingMethods)), SyntaxKind.InvocationExpression);
+                ctxt.RegisterOperationAction(Utils.DebuggableWrapper(c => this.AnalyzeInvocation(c, mainThreadAssertingMethods)), OperationKind.Invocation);
             });
         }
 
-        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context, ImmutableArray<CommonInterest.QualifiedMember> mainThreadAssertingMethods)
+        private void AnalyzeInvocation(OperationAnalysisContext context, ImmutableArray<CommonInterest.QualifiedMember> mainThreadAssertingMethods)
         {
-            var functionInfo = Utils.GetContainingFunction((CSharpSyntaxNode)context.Node);
-            if (functionInfo.Function == null)
+            if (!(Utils.GetContainingFunction(context.Operation, context.ContainingSymbol) is IMethodSymbol methodSymbol))
```

Comment:
Why do we need to invoke `Utils.GetContainingFunction` here?
```suggestion
            if (!(context.ContainingSymbol is IMethodSymbol methodSymbol))
```

Reply:
Because the containing symbol is always the method that is a member of the containing type. But `GetContainingFunction` returns the ... containing function, which might be an inline function, lambda, or delegate.

Reply:
At least, I _think_ that's the case.